Sources/DistributedActors/Pattern/WorkerPool.swift

```diff
@@ -44,19 +54,79 @@ public struct WorkerPool<Message: ActorMessage> {
     // MARK: State shared across all states of the WorkerPool
 
-    var settings: WorkerPoolSettings<Message>
+    var settings: WorkerPoolSettings<Worker>
     var selector: Selector {
         self.settings.selector
     }
-
-    init(settings: WorkerPoolSettings<Message>) {
+    
+    /// The worker pool. Worker will be selected round-robin.
+    private var workers: [Weak<Worker>] = []
+    private var roundRobinPos = 0
+    
+    /// Boolean flag to help determine if pool becomes empty because at least one worker has terminated.
+    private var hasTerminatedWorkers = false
+
+    init(settings: WorkerPoolSettings<Worker>, system: ActorSystem) async {
         self.settings = settings
+        self.actorSystem = system
+
+        switch self.selector {
+        case .dynamic(let key):
+            Task {
+                for await worker in await self.actorSystem.receptionist.subscribe(to: key) {
+                    system.log.log(level: self.settings.logLevel, "Got listing member for \(self.selector): \(worker)")
+                    self.workers.append(Weak(worker))
+                    watchTermination(of: worker) { self.onWorkerTerminated(id: $0) }
+                }
+            }
+        case .static(let workers):
+            self.workers.append(contentsOf: workers.map(Weak.init))
+            workers.forEach { worker in
+                watchTermination(of: worker) { self.onWorkerTerminated(id: $0) }
+            }
+        }
+    }
+    
+    public distributed func submit(work: WorkItem) async throws -> WorkResult {
+        let worker = try await self.selectWorker()
+        return try await worker.submit(work: work)
+    }
+    
+    private func selectWorker() async throws -> Worker {
+        // Wait if we haven't received the initial workers listing yet.
+        // Otherwise, the pool has become empty because all workers have been terminated,
+        // in which case we either wait for new workers or throw error.
+        while self.workers.isEmpty {
+            switch (self.hasTerminatedWorkers, self.settings.whenAllWorkersTerminated) {
+            case (false, _), (true, .awaitNewWorkers):
+                try await Task.sleep(nanoseconds: 100)
+            case (true, .crash(let error)):
+                throw error
+            }
+        }
+
+        if let worker = self.workers[self.roundRobinPos].actor {
+            self.roundRobinPos = (self.roundRobinPos + 1) % self.workers.count
+            return worker
+        } else {
+            // Clean up and try again
+            self.hasTerminatedWorkers = true
+            self.workers.removeAll { $0.actor == nil }
+            self.roundRobinPos = 0
+            return try await self.selectWorker()
+        }
+    }
+    
+    private func onWorkerTerminated(id: ID)  {
+        for worker in self.workers where worker.actor?.id == id {
+            worker.actor = nil
```

Comment:
this should not be necessary since the terminated can only ever happen if the actort was released.
If it was released -> our Weak<> no longer holds it as well; We should though remove the specific actor from our `workers` then though.

I'd suggest moving workers to `[Worker.ID: Weak<Worker>]` this way we can then immediately `workers.remove(forKey: id)` 👍

Reply:
Is it redundant to watchTermination?

Reply:
Amended https://github.com/apple/swift-distributed-actors/pull/912/commits/c987ebe92858fefe6b37f91e1011e9ca065153ba

Reply:
So `watchTermination` is proactive -- you'll get notified in the moment the actor is deinit/terminated - so we can reap them eagerly. 

Also, for remote references, we need the watch since the refs won't just be deallocated